File: frontend/src/services/cowService.ts
import type { Cow } from "../types/cow";
import { apiFetch } from "../lib/api";

export type CreateCowInput = {
  tagNumber: string;
  ownerName: string;
  livestockGroup: string;
  breed: string;
  sex: string;
  healthStatus: string;
  heatStatus: string | null;
  pregnancyStatus: string | null;
  hasCalf: boolean;
  dateOfBirth: string | null;
  purchaseDate: string | null;
  saleDate: string | null;
  purchasePrice: number | null;
  salePrice: number | null;
  notes: string | null;
};

export async function getCows(): Promise<Cow[]> {
  const response = await apiFetch("/cows");
  return response.json();
}

export async function createCow(cowData: CreateCowInput): Promise<Cow> {
  const response = await apiFetch("/cows", {
    method: "POST",
    body: JSON.stringify({
      tagNumber: cowData.tagNumber,
      ownerName: cowData.ownerName,
      livestockGroup: cowData.livestockGroup,
      sex: cowData.sex,
      breed: cowData.breed,
      dateOfBirth: cowData.dateOfBirth,
      healthStatus: cowData.healthStatus,
      heatStatus: cowData.heatStatus,
      pregnancyStatus: cowData.pregnancyStatus,
      hasCalf: cowData.hasCalf,
      purchasePrice: cowData.purchasePrice,
      salePrice: cowData.salePrice,
      purchaseDate: cowData.purchaseDate,
      saleDate: cowData.saleDate,
      notes: cowData.notes,
    }),
  });

  return response.json();
}

export async function getCowById(id: string): Promise<Cow> {
  const response = await apiFetch(`/cows/${id}`);
  return response.json();
}

export async function archiveCow(id: string): Promise<void> {
  await apiFetch(`/cows/${id}/archive`, {
    method: "PUT",
  });
}

export async function deleteCow(id: string): Promise<void> {
  await apiFetch(`/cows/${id}`, {
    method: "DELETE",
  });
}
export async function getRemovedCows(): Promise<Cow[]> {
  const response = await apiFetch("/cows/removed");
  return response.json();
}

export async function restoreCow(id: string): Promise<void> {
  await apiFetch(`/cows/${id}/restore`, {
    method: "PUT",
  });
}

export async function updateCow(
  id: string,
  cowData: CreateCowInput,
): Promise<Cow> {
  const response = await apiFetch(`/cows/${id}`, {
    method: "PUT",
    body: JSON.stringify({
      tagNumber: cowData.tagNumber,
      ownerName: cowData.ownerName,
      livestockGroup: cowData.livestockGroup,
      sex: cowData.sex,
      breed: cowData.breed,
      dateOfBirth: cowData.dateOfBirth,
      healthStatus: cowData.healthStatus,
      heatStatus: cowData.heatStatus,
      pregnancyStatus: cowData.pregnancyStatus,
      hasCalf: cowData.hasCalf,
      purchasePrice: cowData.purchasePrice,
      salePrice: cowData.salePrice,
      purchaseDate: cowData.purchaseDate,
      saleDate: cowData.saleDate,
      notes: cowData.notes,
    }),
  });

  return response.json();
}

File: backend/HerdFlow.Api/Services/CowService.cs
using HerdFlow.Api.DTOs;
using HerdFlow.Api.Models;
using HerdFlow.Api.Data;

namespace HerdFlow.Api.Services;

public class CowService
{
    private readonly AppDbContext _context;

    public CowService(AppDbContext context)
    {
        _context = context;
    }
    public List<Cow> GetCows()
    {
        return _context.Cows.ToList();
    }
    public Cow? GetCowById(int id)
    {
        return _context.Cows.FirstOrDefault(c => c.Id == id);
    }
    public Cow? UpdateCow(int id, CreateCowDto dto)
    {
        var cow = _context.Cows.FirstOrDefault(c => c.Id == id);

        if (cow == null)
            return null;

        cow.TagNumber = dto.TagNumber;
        cow.OwnerName = dto.OwnerName;
        cow.LivestockGroup = dto.LivestockGroup;
        cow.Breed = dto.Breed;
        cow.HealthStatus = dto.HealthStatus;
        cow.BreedingStatus = dto.BreedingStatus;
        cow.Sex = dto.Sex;
        cow.DateOfBirth = dto.DateOfBirth;
        cow.Notes = dto.Notes;

        cow.PurchasePrice = dto.PurchasePrice;
        cow.SalePrice = dto.SalePrice;
        cow.PurchaseDate = dto.PurchaseDate;
        cow.SaleDate = dto.SaleDate;

        _context.SaveChanges();

        return cow;
    }
    public Cow CreateCow(CreateCowDto dto)
    {
        var cow = new Cow
        {
            TagNumber = dto.TagNumber,
            OwnerName = dto.OwnerName,
            LivestockGroup = dto.LivestockGroup,
            Sex = dto.Sex,
            Breed = dto.Breed,
            DateOfBirth = dto.DateOfBirth,
            HealthStatus = dto.HealthStatus,
            HeatStatus = dto.HeatStatus,
            BreedingStatus = dto.BreedingStatus,
            PurchasePrice = dto.PurchasePrice,
            SalePrice = dto.SalePrice,
            PurchaseDate = dto.PurchaseDate,
            SaleDate = dto.SaleDate,
            Notes = dto.Notes,
            HarvestDate = dto.HarvestDate
        };
        _context.Cows.Add(cow);
        _context.SaveChanges();
        return cow;
    }
    public bool DeleteCow(int id)
    {
        var cow = _context.Cows.FirstOrDefault(c => c.Id == id);

        if (cow == null)
            return false;

        _context.Cows.Remove(cow);
        _context.SaveChanges();

        return true;
    }
}
